go-controller/pkg/ovnwebhook/nodeadmission.go

```diff
@@ -0,0 +1,191 @@
+	util.OvnNodeChassisID: func(v annotationChange, nodeName string) error {
+		if v.action == removed {
+			return fmt.Errorf("%s cannot be removed", util.OvnNodeChassisID)
+		}
+		if v.action == changed {
+			return fmt.Errorf("%s cannot be changed once set", util.OvnNodeChassisID)
+		}
+		return nil
+	},
+	util.OvnNodeZoneName: func(v annotationChange, nodeName string) error {
+		// it is allowed for the annotation to be set to "global" or <nodeName> initially
+		if (v.action == added || v.action == changed) &&
+			(v.value == types.OvnDefaultZone || v.value == nodeName) {
+			return nil
+		}
+
+		return fmt.Errorf("%s can only be set to %s or %s, it cannot be removed", util.OvnNodeZoneName, types.OvnDefaultZone, nodeName)
+	},
+}
+
+// interconnectNodeAnnotationChecks holds annotations allowed for ovnkube-node:<nodeName> users in IC environments
+var interconnectNodeAnnotationChecks = map[string]checkNodeAnnot{
+	util.OvnNodeMigratedZoneName: func(v annotationChange, nodeName string) error {
+		// it is allowed for the annotation to be set to <nodeName>
+		if (v.action == added || v.action == changed) && v.value == nodeName {
+			return nil
+		}
+
+		return fmt.Errorf("%s can only be set to %s, it cannot be removed", util.OvnNodeMigratedZoneName, nodeName)
+	},
+}
+
+// hybridOverlayNodeAnnotationChecks holds annotations allowed for ovnkube-node:<nodeName> users hybrid overlay environments
+var hybridOverlayNodeAnnotationChecks = map[string]checkNodeAnnot{
+	hotypes.HybridOverlayDRMAC: nil,
+	hotypes.HybridOverlayDRIP:  nil,
+}
+
+type NodeAdmission struct {
+	annotationChecks  map[string]checkNodeAnnot
+	annotationKeys    sets.Set[string]
+	extraAllowedUsers sets.Set[string]
+}
+
+func NewNodeAdmissionWebhook(enableInterconnect, enableHybridOverlay bool, extraAllowedUsers ...string) *NodeAdmission {
+	checks := make(map[string]checkNodeAnnot)
+	maps.Copy(checks, commonNodeAnnotationChecks)
+	if enableInterconnect {
+		maps.Copy(checks, interconnectNodeAnnotationChecks)
+	}
+	if enableHybridOverlay {
+		maps.Copy(checks, hybridOverlayNodeAnnotationChecks)
+	}
+	return &NodeAdmission{
+		annotationChecks:  checks,
+		annotationKeys:    sets.New[string](maps.Keys(checks)...),
+		extraAllowedUsers: sets.New[string](extraAllowedUsers...),
+	}
+}
+
+var _ admission.CustomValidator = &NodeAdmission{}
+
+func (p NodeAdmission) ValidateCreate(_ context.Context, _ runtime.Object) (warnings admission.Warnings, err error) {
+	// Ignore creation, the webhook is configured to only handle nodes/status updates
+	return nil, nil
+}
+
+func (p NodeAdmission) ValidateDelete(_ context.Context, _ runtime.Object) (warnings admission.Warnings, err error) {
+	// Ignore deletion, the webhook is configured to only handle nodes/status updates
+	return nil, nil
+}
+
+func (p NodeAdmission) ValidateUpdate(ctx context.Context, oldObj, newObj runtime.Object) (warnings admission.Warnings, err error) {
+	oldNode := oldObj.(*corev1.Node)
```

Comment:
again stupid question: why do we care only about updates and not creates/deletes?

Reply:
In this effort we want to ensure that a particular ovnkube-node is not able to affect workloads running on different nodes. Since ovnkube-node is not allowed to create/delete pods/nodes this is not something I want to handle in this PR. We have discussed building on top of this effort in the future to handle more cases that are not directly caused by our components.